akka-projection-slick/src/main/scala/akka/projection/slick/internal/SlickOffsetStore.scala

```diff
@@ -23,39 +24,73 @@ import slick.jdbc.JdbcProfile
     val db: P#Backend#Database,
     val profile: P,
     clock: Clock) {
+  import OffsetSerialization.MultipleOffsets
+  import OffsetSerialization.SingleOffset
   import OffsetSerialization.fromStorageRepresentation
   import OffsetSerialization.toStorageRepresentation
   import profile.api._
 
-  def this(db: P#Backend#Database, profile: P) = this(db, profile, Clock.systemUTC())
+  def this(db: P#Backend#Database, profile: P) =
+    this(db, profile, Clock.systemUTC())
 
   def readOffset[Offset](projectionId: ProjectionId)(implicit ec: ExecutionContext): Future[Option[Offset]] = {
     val action =
-      offsetTable.filter(_.projectionId === projectionId.id).result.headOption.map { maybeRow =>
-        maybeRow.map(row => fromStorageRepresentation[Offset](row.offsetStr, row.manifest))
+      offsetTable.filter(_.projectionName === projectionId.name).result.map { maybeRow =>
+        maybeRow.map(row => SingleOffset(projectionId, row.manifest, row.offsetStr, row.mergeable))
       }
```

Comment:
I think this is a good first implementation. Later we could optimize for the single-offset case by first reading one row for the `projectionId.id`, and if that is mergable == false then it's know to be single-offset, otherwise (including when not found) we can read all rows for the `projectionId.name`.

Reply:
Yes.  I think the worst case would be loading all offsets for a Kafka consumer group with 1000's of partitions, but even then the `readOffset` use case only happens at startup or a rebalance, so it's going to be a negligible performance impact in the majority of use cases.